### ISSUE TITLE
'nuts': Convert European Regional Data

### ISSUE THREAD

**Issue by AAoritz** [ropensci/software-review#623]
**Date accepted:** 2024-03-14
Submitting Author Name: Moritz Hennicke
Submitting Author Github Handle: @AAoritz
Other Package Authors Github handles: @krausewe
Repository:  https://github.com/AAoritz/nuts/
Version submitted: 0.0.0.9000
Submission type: Standard
Editor: @maelle
Reviewers: @nolwenn, @jospueyo

Archive: TBD
Version accepted: TBD
Language: en

---

-   Paste the full DESCRIPTION file inside a code block below:

```
Package: nuts
Title: 'nuts': Convert European Regional Data
Version: 0.0.0.9000
Authors@R: c(person("Moritz", "Hennicke", email = "AAoritz@posteo.de", role = c("aut", "cre"),
		   comment = c(ORCID = "0000-0001-6811-1821")),
	     person("Werner", "Krause", email = "werner.krause@uni-potsdam.de", role = c("aut"),
		   comment = c(ORCID = "0000-0002-5069-7964")))
Description: Motivated by changing administrative boundaries over time, the 'nuts' package can convert European regional data with NUTS codes between versions (2006, 2010, 2013, 2016 and 2021) and levels (NUTS 1, NUTS 2 and NUTS 3). The package uses spatial interpolation based on granular (100mx100m) population and land use data provided by the European Commission's Joint Research Center.  
URL: https://AAoritz.github.io/nuts/
BugReports: https://github.com/AAoritz/nuts/
License: MIT + file LICENSE
Encoding: UTF-8
Roxygen: list(markdown = TRUE)
RoxygenNote: 7.3.1
Suggests: 
    distill,
    eurostat,
    formatR,
    ggalluvial,
    ggfittext,
    ggplot2,
    ggpubr,
    ggrepel,
    gridExtra,
    kableExtra,
    knitr,
    raster,
    RColorBrewer,
    readr,
    rmarkdown,
    sf,
    terra,
    testthat,
    tidyr
Config/testthat/edition: 3
VignetteBuilder: knitr
Imports:
	crayon,
	dplyr,
	stringr
Depends: 
    R (>= 3.5.0)
LazyData: true

```

## Scope

- Please indicate which category or categories from our [package fit policies](https://ropensci.github.io/dev_guide/policies.html#package-categories) this package falls under: (Please check an appropriate box below. If you are unsure, we suggest you make a pre-submission inquiry.):

	- [ ] data retrieval
	- [ ] data extraction
	- [X] data munging
	- [ ] data deposition
	- [X] data validation and testing
	- [ ] workflow automation
	- [ ] version control
	- [ ] citation management and bibliometrics
	- [ ] scientific software wrappers
	- [ ] field and lab reproducibility tools
	- [ ] database software bindings
	- [X] geospatial data
	- [ ] text analysis

- Explain how and why the package falls under these categories (briefly, 1-2 sentences):

Data munging: Linking regional data from different sources at the level of NUTS codes is often complicated by the usage of different versions or varying levels of geographical granularity across sources. The package can be used to harmonize NUTS versions and levels across different data sources using spatial interpolation. 

Data validation and testing: The package includes routine tasks to test for the validity and completeness of NUTS codes. 

Geospatial data: NUTS codes are the dominant format for European regional data.

-   Who is the target audience and what are scientific applications of this package?

The target audience are academics, journalists and data scientists interested in European regional data. Users who want to exploit changes within NUTS regions over time face the challenge that administrative boundaries are redrawn over time. The package enables the construction of consistent panel data across NUTS regions and over time through the harmonization of NUTS regions to one common version or level of granularity. 

-   Are there other R packages that accomplish the same thing? If so, how does yours differ or meet [our criteria for best-in-category](https://ropensci.github.io/dev_guide/policies.html#overlap)?

To our knowledge there is currently no package that is targeted at the conversion of NUTS versions using spatial interpolation. 

The `regions` package allows for re-coding of NUTS codes from version to version without spatial interpolation. The package offers some code validation routines, but not an automated detection of the NUTS version used.    

-   (If applicable) Does your package comply with our [guidance around _Ethics, Data Privacy and Human Subjects Research_](https://devguide.ropensci.org/policies.html#ethics-data-privacy-and-human-subjects-research)?

Yes

-   If you made a pre-submission inquiry, please paste the link to the corresponding issue, forum post, or other discussion, or @tag the editor you contacted.

-   Explain reasons for any [`pkgcheck` items](https://docs.ropensci.org/pkgcheck/) which your package is unable to pass.

## Technical checks

Confirm each of the following by checking the box.

- [X] I have read the [rOpenSci packaging guide](https://devguide.ropensci.org/building.html).
- [X] I have read the [author guide](https://devdevguide.netlify.app/authors-guide.html) and I expect to maintain this package for at least 2 years or to find a replacement.

This package:

- [X] does not violate the Terms of Service of any service it interacts with.
- [X] has a CRAN and OSI accepted license.
- [X] contains a [README with instructions for installing the development version](https://ropensci.github.io/dev_guide/building.html#readme).
- [X] includes [documentation with examples for all functions, created with roxygen2](https://ropensci.github.io/dev_guide/building.html#documentation).
- [X] contains a vignette with examples of its essential functions and uses.
- [X] has a [test suite](https://ropensci.github.io/dev_guide/building.html#testing).
- [X] has [continuous integration](https://ropensci.github.io/dev_guide/ci.html), including reporting of test coverage.

## Publication options

- [X] Do you intend for this package to go on CRAN?

## Code of conduct

- [X] I agree to abide by [rOpenSci's Code of Conduct](https://ropensci.org/code-of-conduct/) during the review process and in maintaining my package should it be accepted.

**Comment by AAoritz** [comment]
Hi rOpenSci Team and @ropensci-review-bot!

- We added a contributing file with this [commit](https://github.com/AAoritz/nuts/commit/d3c15290d28e1dd4e3d4e4a99c25d45bc831b8f6)
- The package uses an HTML vignette based on [`distill`](https://rstudio.github.io/distill/). The bot prefers in [`check-vignette.R`](https://github.com/ropensci-review-tools/pkgcheck/blob/e499f62755a76c237a50e26ec77d3bc695ace2a0/R/check-vignette.R#L58) `rmarkdown` HTML files. Should we switch to `rmarkdown`?

Thank your for your time and consideration!

**Comment by mpadge** [comment]
@AAoritz Sorry about the inconvenience there. We've updated our system to detect distill vignettes, so your package will now pass those tests. (The server might take a few days to incorporate those updates; please be paitent.)

**Comment by jhollist** [comment]
Yep, thanks @mpadge.  Will wait a bit of sending another check request to let those changes work through.

@AAoritz I think this looks good and is a fit for rOpenSci.  I will start working on finding a handling editor for your submission!

**Comment by AAoritz** [comment]
Great news @jhollist! Thank you!

**Comment by jhollist** [comment]
@ropensci-review-bot check package

**Comment by jhollist** [comment]
@AAoritz looking good!  Assigning editor in just a few.

**Comment by maelle** [comment]
# Editor checks:

- [x] **Documentation**: The package has sufficient documentation available online (README, pkgdown docs) to allow for an assessment of functionality and scope without installing the package. In particular,
    - [x] Is the case for the package well made?
    - [ ] Is the reference index page clear (grouped by topic if necessary)?
    - [x] Are vignettes readable, sufficiently detailed and not just perfunctory?
- [x] **Fit**: The package meets criteria for [fit](https://devguide.ropensci.org/policies.html#package-categories) and [overlap](https://devguide.ropensci.org/policies.html#overlap).
- [ ] **Installation instructions:** Are installation instructions clear enough for human users?
- [x] **Tests**: If the package has some interactivity / HTTP / plot production etc. are the tests using [state-of-the-art tooling](https://devguide.ropensci.org/building.html#testing)?
- [ ] **Contributing information**: Is the documentation for contribution clear enough e.g. tokens for tests, playgrounds?
- [x] **License:** The package has a CRAN or OSI accepted license.
- [x] **Project management**: Are the issue and PR trackers in a good shape, e.g. are there outstanding bugs, is it clear when feature requests are meant to be tackled?
---

# Editor comments

Many thanks for your submission @AAoritz & @krausewe! Useful package! I know of the French equivalent to nuts, COGjugaison, so was vaguely aware of the idea https://antuki.github.io/COGugaison/ :smile_cat: 

I have a few comments that I'd like to resolve or discuss before I proceed to the reviewer search.

## Contributing guide

- In the `CONTRIBUTING.md` file could you please remove the tidyverse specific bits such as " For more detailed info about contributing to this, and other tidyverse packages,"?

- I was surprised to see optional dependencies before Imports in `DESCRIPTION`. Would you mind running `desc::desc_normalize()` and see whether you like the result?

## Docs

- Instead of remotes in the README you could recommend pak, `pak::pak("AAoritz/nuts")` but in any case I don't think the `force = TRUE` argument is warranted, is it?

- Since there is only one vignette I'd recommend naming it "nuts.Rmd" so that it is automatically added in the pkgdown website navbar as "Get started". See https://pkgdown.r-lib.org/reference/build_articles.html#get-started

- Your images have alternative text but that text is not complete, it does not describe the image so that a screenreader user would miss a lot. https://www.w3.org/WAI/tutorials/images/informative/

- I was surprised by the short pipeline in https://github.com/AAoritz/nuts/blob/cf771ced02a20b226b56e6908959044ca62804e4/vignettes/nuts-vignette.Rmd#L247, could it avoid using the pipe, or if it does, could `filter(` go on a separate line? I feel it's an unusual pattern (but I might be wrong). I see this pattern of one-line pipelines in the tests too so it might be a personal preference! It's more noticeable/important in the user facing code but I won't impose _my_ personal style.

- The reference index is short but it might make sense to group the functions then the datasets. https://pkgdown.r-lib.org/reference/build_reference.html#reference-index

## User interface

- The function output of https://aaoritz.github.io/nuts/reference/classify_nuts.html#value is a list where you identify parts via their index. Could you add names to the output? Something like

```r
output <- list(
  data = data, 
  versions_data = data_all_versions,  
  missing_data = data_missing_nuts
)
```

so that then the user could use `output [["missing_data"]]` instead of `output[[3]]` which is less readable. 

- Regarding the messages, 
    - I see you use crayon for information but `stop()` for error messages. The crayon package has been superseeded by cli (see https://cli.r-lib.org/ and https://blog.r-hub.io/2023/11/30/cliff-notes-about-cli/), according to https://cran.r-project.org/web/packages/crayon/index.html. You could keep using crayon (it's still supported) but it might make sense switching, as cli provides really nice features. Code like
    
```r
stop("Input 'nuts_code' must be provided as a string.")
```

could become

```r
cli::cli_abort("Input {.arg nuts_code} must be provided as a string, not {.obj_type_friendly {nuts_code}}.")
```

(note that this function also requires importing rlang in DESCRIPTION)

and info messages could use `cli_alert_` functions https://cli.r-lib.org/reference/cli_alert.html 

    - Most importantly, at the moment there is no way for the user to turn off messages. This is guidance we've just added to the development version of the dev guide so pretty new but I would still like this to be implemented as it might improve the quality of life of the users (including you, for instance in tests). See https://devdevguide.netlify.app/pkg_building#console-messages, https://blog.r-hub.io/2023/11/30/cliff-notes-about-cli/#how-to-make-cli-quiet-or-not if you use cli, and, hum, a draft blog post @mpadge and I are working on https://github.com/ropensci/roweb3/pull/717 Based on the interface tooling you end up selecting I am happy to provide tips about silencing if needed. Thanks for your understanding. :sweat_smile: 

## Code

- Why does the package import whole packages rather than individual functions? Furthermore the imports should happen only once so they could happen in the package-level manual page source. You can create it with https://usethis.r-lib.org/reference/use_package_doc.html whose docs say "This .R file is also a good place for roxygen directives that apply to the whole package (vs. a specific function), such as global namespace tags like @importFrom."

- In lines such as https://github.com/AAoritz/nuts/blob/cf771ced02a20b226b56e6908959044ca62804e4/R/convert_nuts_version.R#L128 the code might benefit from the use of ["explaining variables"](https://blog.thepete.net/blog/2021/06/24/explaining-variable/), for instance (with a better name probably, again an example :sweat_smile: )

```r
not_enough_codes <- (length(data$from_code[check_nuts_codes]) < length(data$from_code) &&
               length(data$from_code[check_nuts_codes]) > 0)
if  (not_enough_codes)
```
(oh and the use of && instead of & see https://lintr.r-lib.org/reference/vector_logic_linter.html)

## Tests

- You do not need `data()` to load package data in tests.

- I'd recommend not having top-level code in tests which means no code outside of `test_that(`. This makes tests more self contained. So in a helper file, `tests/testthat/helper-test-data.R` for instance you'd have the function

```r
manure_indic_DE_2003 <- function() {
manure %>%
  filter(nchar(geo) == 4) %>%
  filter(indic_ag == "I07A_EQ_Y") %>%
  select(-indic_ag) %>%
  filter(grepl("^DE", geo)) %>%
  filter(time == 2003) %>%
  select(-time)
}
```

and in `tests/testthat/test-classify_nuts.R` you'd have

```r
test_that("Needs geo var 1", {
  expect_error(
    manure_indic_DE_2003() %>% classify_nuts(nuts_code = NULL),
    "Input 'nuts_code' cannot be NULL."
  )
})

test_that("Needs geo var 2", {
  expect_error(manure_indic_DE_2003() %>% classify_nuts())
})

test_that("nuts_code not valid", {
  expect_error(
    manure_indic_DE_2003() %>% classify_nuts(nuts_code = 1),
    "Input 'nuts_code' must be provided as a string."
  )
})
```

See https://blog.r-hub.io/2020/11/18/testthat-utility-belt/ and https://r-pkgs.org/testing-design.html + https://r-pkgs.org/testing-advanced.html

- Instead of using `set.seed()`, in tests please use the withr version: https://withr.r-lib.org/reference/with_seed.html
---

Happy to answer any question and to discuss! Thanks again for submitting your package!

**Comment by AAoritz** [comment]
Hi @maelle ! Many thanks for these super useful checks and comments. @krausewe and I start looking into them now and we will get back to you soon.

**Comment by krausewe** [comment]
Hi @maelle!

We have addressed the very helpful remarks & recommendations. 

Please find below the list of commits.

We are looking forward to the next steps!

# Contributing guide
- Remove tidyverse specific points in `CONTRIBUTING`: https://github.com/AAoritz/nuts/commit/f5c4a5a4049c52c9f34dd253d99a0eb2515af164
- Normalize `DESCRIPTION`: https://github.com/AAoritz/nuts/commit/b80c8db0ad253a631b6b06f8fa20a61523c0d218

# Docs
- Change install to `pak` in `README`: https://github.com/AAoritz/nuts/commit/51c2e44addb5b5052421b967cb9ddc11837fb6af
- Rename vignette to `nuts.Rmd`: https://github.com/AAoritz/nuts/commit/3fe96c87ece74a6e28d35f350302d282523238e8
- Include alt-text of images in vignette: https://github.com/AAoritz/nuts/commit/51f1cb3c67b1e85acaef84dc3937620c963a3985
- Add line breaks for one-line pipelines: https://github.com/AAoritz/nuts/commit/baf01ff9227695fa60e95cc6ee61c7d6c3189202
- Group functions and data `pkgdown`: https://github.com/AAoritz/nuts/commit/f11a08070d0c1356339936c742165cfe6a551c34

# User interface
- Simplify names of output list of `classify`: https://github.com/AAoritz/nuts/commit/a5bbea99ace46bd32be7bc0628d239ebe0f319f4
- Use `cli` messages: https://github.com/AAoritz/nuts/commit/8cb9be6e0fa526f1fc43cbaeb8ab5b3bec89c8e8
- Silencing messages: https://github.com/AAoritz/nuts/commit/9a93ff174389a922ca7bae39daa039d93eb7454b

# Code
- Import functions as opposed to entire packages: https://github.com/AAoritz/nuts/commit/d2744d97cd43dd7a45c52416640633cfd604d34f
- Use explaining variables and double `&&`: https://github.com/AAoritz/nuts/commit/4bc0303f8b2be824ac554d31cb772ac27d242140

# Tests
- Don't load `data()`: https://github.com/AAoritz/nuts/commit/e2ba58f62169b8c84b33817faf305dfc6aeb6c09 
- Include helper file that prepares data: https://github.com/AAoritz/nuts/commit/e2ba58f62169b8c84b33817faf305dfc6aeb6c09 
- Replace `set.seed()` with `withr`: https://github.com/AAoritz/nuts/commit/0d9e3ad4084780c69179f52afd21c30b1eedba6c

**Comment by maelle** [comment]
Awesome, thank you! I'll now look for reviewers.

Note that @mpadge's and my tech note is now published: https://ropensci.org/blog/2024/02/06/verbosity-control-packages/

**Comment by krausewe** [comment]
@ropensci-review-bot. 

Added Peer Review Status.

`NEWS.md` was already created and included.

**Comment by maelle** [comment]
@ropensci-review-bot add @nolwenn to reviewers

**Comment by AAoritz** [comment]
Hi @maelle! We are a bit lost with implementing verbosity control with `local_options`. This example from your blog does not seem to work with `cli`:

```
> pkg_message <- function(...) {
+   is_verbose_mode <- (getOption("mypackage.verbose", "quiet") == "verbose")
+   if (is_verbose_mode) {
+     # Options local to this function only; reset on exit!
+     rlang::local_options(rlib_message_verbosity = "verbose")
+   }
+   rlang::inform(...)
+ }
> pkg_message("normal message")
normal message
> rlang::local_options(rlib_message_verbosity = "quiet")
Setting global deferred event(s).
i These will be run:
  * Automatically, when the R session ends.
  * On demand, if you call `withr::deferred_run()`.
i Use `withr::deferred_clear()` to clear them without executing.
> pkg_message("suppressed message")
> rlang::local_options(mypackage.verbose = "verbose")
> pkg_message("reawakened message")
reawakened message


> withr::deferred_run()
Ran 2/2 deferred expressions
> pkg_message <- function(...) {
+   is_verbose_mode <- (getOption("mypackage.verbose", "quiet") == "verbose")
+   if (is_verbose_mode) {
+     # Options local to this function only; reset on exit!
+     rlang::local_options(rlib_message_verbosity = "verbose")
+   }
+   cli::cli_h1(...)
+ }
> pkg_message("normal message")

── normal message ───────────────────────────────────────────────────────────────────────────────────────────────
> rlang::local_options(rlib_message_verbosity = "quiet")
Setting global deferred event(s).
i These will be run:
  * Automatically, when the R session ends.
  * On demand, if you call `withr::deferred_run()`.
i Use `withr::deferred_clear()` to clear them without executing.
> pkg_message("suppressed message")

── suppressed message ───────────────────────────────────────────────────────────────────────────────────────────
> rlang::local_options(mypackage.verbose = "verbose")
> pkg_message("reawakened message")

── reawakened message ───────────────────────────────────────────────────────────────────────────────────────────
> 
```

**Comment by maelle** [comment]
Right, it seems specific to rlang (cc @mpadge). 

How about something simpler like

``` r
pkg_message <- function(...) {
  is_verbose_mode <- (getOption("mypackage.verbose", "quiet") == "verbose")
  if (is_verbose_mode) {
    cli::cli_alert(...)
  }
  
}

withr::with_options(list("mypackage.verbose" = "quiet"), {
  pkg_message("pof")
})

withr::with_options(list("mypackage.verbose" = "verbose"), {
  pkg_message("pof")
})
#> → pof
```

<sup>Created on 2024-02-13 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

:thinking:

**Comment by AAoritz** [comment]
Thank you @maelle and @mpadge! I wrapped our `cli` message as suggested in https://github.com/AAoritz/nuts/commit/47ee441ea6ea98b30326157b803f81dba5557b42 by @maelle. Here is the result: 

``` r
rlang::local_options(nuts.verbose = "quiet")
df <- patents %>%
  filter(unit == "NR", nchar(geo) == 4, time == 2012) %>%
  filter(grepl("^DE", geo)) %>%
  classify_nuts(data = ., nuts_code = "geo")

withr::deferred_run()
#> Ran 1/1 deferred expressions
rlang::local_options(nuts.verbose = "verbose")
df <- patents %>%
  filter(unit == "NR", nchar(geo) == 4, time == 2012) %>%
  filter(grepl("^DE", geo)) %>%
  classify_nuts(data = ., nuts_code = "geo")
#> 
#> ── Classifying version of NUTS codes ───────────────────────────────────────────
#> Within groups defined by country:
#> ! These NUTS codes cannot be identified or classified: DEXX and DEZZ.
#> ✔ Unique NUTS version classified.
#> ✔ No missing NUTS codes.

withr::deferred_run()
#> Ran 2/2 deferred expressions
rlang::local_options(rlib_message_verbosity = "quiet")
df <- patents %>%
  filter(unit == "NR", nchar(geo) == 4, time == 2012) %>%
  filter(grepl("^DE", geo)) %>%
  classify_nuts(data = ., nuts_code = "geo")
```

<sup>Created on 2024-02-14 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

How would you recommend to document this option? Below the examples in the function level help/documentation?

**Comment by mpadge** [comment]
@AAoritz Would you please be so kind as to ask that question on [our discussion forum](https://discuss.ropensci.org/t/ropensci-please-shut-up-verbosity-control-in-packages/3790/12)? The issue of where to document verbosity control is important, and would be good to have in that general context, rather than here.

**Comment by maelle** [comment]
@ropensci-review-bot add @jospueyo to reviewers

**Comment by jospueyo** [comment]
## Package Review

-   **Briefly describe any working relationship you have (had) with the package authors.**
-   [x] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html#coi) for me to review this work.

#### Documentation

The package includes all the following forms of documentation:

-   [x] **A statement of need:** clearly stating problems the software is designed to solve and its target audience in README
-   [x] **Installation instructions:** for the development version of package and any non-standard dependencies in README
-   [x] **Vignette(s):** demonstrating major functionality that runs successfully locally
-   [x] **Function Documentation:** for all exported functions
-   [x] **Examples:** (that run successfully locally) for all exported functions
-   [x] **Community guidelines:** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

#### Functionality

-   [x] **Installation:** Installation succeeds as documented.
-   [ ] **Functionality:** Any functional claims of the software have been confirmed.
-   [x] **Performance:** Any performance claims of the software have been confirmed.
-   [x] **Automated tests:** Unit tests cover essential functions of the package and a reasonable range of inputs and conditions. All tests pass on the local machine.
-   [x] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines.

Estimated hours spent reviewing:

-   [x] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer (“rev” role) in the package DESCRIPTION file.

------------------------------------------------------------------------

### Review Comments

The package is useful, very useful indeed, and also well designed. Most of my comments below are just suggestions that IMHO will make the package more intuitive, more robust, or more easily maintainable. Don't take them as mandatory. At the end, this is your package and you should have the last word.

Disclaimer: Apologies if some comments sound though or rude. English is not my native language.

#### General comments

Function naming does not follow rOpenSci package guidelines. Specifically, the design `object_verb()`. Please, start your exported functions with `nuts_`, this also facilitates autocompletion.

Variables in vignettes should use underscores and not points, that can clash with S3 methods.

`group_vars` argument could be changed to `.by`, to be consinstent with `dplyr` functions.

`convert_nuts_levels` might be changed to `nuts_aggregate` to be explicit about that the level change is only aggregation and not disaggregation.

Provide wrapper functions to inspect sub-elements of `classifiy_nuts`, something like: 

- `nuts_get_data()` 
- `nuts_get_version()`
- `nuts_get_missing()`

`missing_rm` could also provide some imputation technique, such as using the mean. In the future, it might be linked to imputation packages such as `mice`, for instance.

There are some inconsistencies in your constructions that can make the package harder to mantain, especially if there is a maintainer change in the future. For instance, sometimes you are explicit inside your conditions: `if(condition == FALSE) do something...` vs `if(!condition) do something...`. You should check the constructions you used and try to be consistent. BTW, you should avoid using `T` and `F`, replace them with `TRUE` and `FALSE`.

You should not supply default values in mandatory arguments. Otherwise, if you run the function without providing these values, it raises a cryptographic error instead of the usual `Error in classify_nuts() : argument "data" is missing, with no default`.

When an argument accepts a set of values, such as`multiple_versions` or `weights`, it’s preferable to pass a vector with all options as default to show all possible values. E.g. `multiple_versions = c('break', 'most_frequent')`. Then, you use the first element, which would be `break` by default. BTW, `break` is a concept that not anybody might be familiar with. Maybe, using `error` would be more intuitive.

When you check if an argument is one of the accepted values, you should contemplate that users could pass a vector of length > 1 (never underestimate end-users). If this happens, an error is given because `if()` can only accept 1-length vectors. You should do something like `if(!all(nuts_code %in% colnames(data)))`.

The “nuts.verbose” options should be mentioned in the help of each function. Otherwise, users can’t know what options they have about verbosity.

#### Documentation

In the table about converting relative values, some `*` are interpreted by markdown as italics. This makes impossible to interpret the weighted average formulas. You should fix it, possibly converting to equations or to code. You will have to try.

#### Imports

All the scoped verbs (`*_at`) were superseeded in current versions of dplyr. Replace with new forms such as `summarize(across())`.

The pipe from `magrittr` and `as_tibble` from `tibble` package can be imported from `dplyr` and you can remove two packages from imports. You will depend on tibble and magrittr anyway, but not directly but trough `dplyr`.

Likewise, if you don’t want to depend on `stringr`, you can easily replace `str_remove` with `gsub()`, passing `""` as replacement; and `str_detect` with `grepl`.

#### `classify_nuts` (suggested name: `nuts_classify`)

-   L310, you should add the class “nuts.classified” instead of overwriting the class “list”: `class(x) <- c("new_class", class(x)`. Otherwise, if the object is no longer a list and only your custom class, this could break some methods for lists and raise unexpected behavior if the output of `classify_nuts` is used by something else than feed your other functions.

#### `convert_nuts_version` (suggested name: `nuts_convert_version`)

-   L72: There is `inherits(object, class)` to check a class. It’s more robust than `class(x) == "your_class"`.

-   L183: You could do the subset just after the `get()`. This way, you save loading the entire table in memory: `cross_walks <- get("cross_walks")[cross_walks$to_version == to_version, ]`.

#### `convert_nuts_level` (suggested name: `nuts_aggregate`)

-   In the vignettes, you explain that only aggregation is possible, which makes sense. However, this is not reflected in the help page. For instance, why `to_level` can be set to 3 if it’s not possible convert from level 1 or 2 to 3? You should add this in the description. My suggestion of name also makes this clearer.

L151: This chunk of code is repeated in `convert_nuts_version`. This is complex enough to avoid repeating yourself. I suggest moving all this chunk to a new function and call this function in both. This makes you code more maintainable.

Looking forward your responses!

**Comment by maelle** [comment]
Thanks a lot @jospueyo! :pray: 

Could you please add the number of hours you spent reviewing near "Estimated hours spent reviewing:" in the template? Thanks!

**Comment by krausewe** [comment]
Thanks so much @jospueyo! All of this is super helpful.

We will start addressing the recommendations asap!

**Comment by maelle** [comment]
@ropensci-review-bot submit review https://github.com/ropensci/software-review/issues/623#issuecomment-1951446662 time 3

**Comment by nolwenn** [comment]
## nuts package review

**Briefly describe any working relationship you have (had) with the package authors.**

-   [x] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html#coi) for me to review this work (if you are unsure whether you are in conflict, please speak to your editor *before* starting your review).

#### Documentation

The package includes all the following forms of documentation:

-   [x] **A statement of need:** clearly stating problems the software is designed to solve and its target audience in README

-   [ ] **Installation instructions:** for the development version of package and any non-standard dependencies in README

-   [ ] **Vignette(s):** demonstrating major functionality that runs successfully locally

-   [x] **Function Documentation:** for all exported functions

-   [x] **Examples:** (that run successfully locally) for all exported functions

-   [x] **Community guidelines:** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

\#### Functionality

-   [ ] **Installation:** Installation succeeds as documented.
-   [x] **Functionality:** Any functional claims of the software been confirmed.
-   [x] **Performance:** Any performance claims of the software been confirmed.
-   [x] **Automated tests:** Unit tests cover essential functions of the package and a reasonable range of inputs and conditions. All tests pass on the local machine.
-   [x] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines.

Estimated hours spent reviewing: 2

-   [x] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

------------------------------------------------------------------------

### Review Comments

The package could be really useful for many applications where spatial-temporal question is involved. At the time of my review, the authors have already considered some of the comments of reviewer 1 so you will excuse me if some of my comments might be outdated.

#### Installation instructions github

you have a error, an extra "/" . It should be `r pak::pak("AAoritz/nuts")` and not `r pak::pak("AAoritz/nuts/")`

At first load I had the message:

Error: package or namespace load failed for ‘nuts’ in loadNamespace(j \<- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]): namespace ‘vctrs’ 0.6.3 is already loaded, but \>= 0.6.4 is required

I had to close and reopen R/Rstudio then it worked fine.

#### Vignette(s)

The Rmd file has been updated with the previous recommendations by @jospueyo concerning the naming of the function (which is good) but not the HTML version <https://aaoritz.github.io/nuts/articles/nuts-vignette.html>. The vignette in HTML version need to be rebuilt and push for online display.

In the vignette, the author did not explicitly call the libraries required for running the data management needed for the examples like the meta-library _tidyverse_ with _dplyr_ for the _filter()_ functions and other data management functions.

In the vignette, maps are displayed with various NUTS over time. Even so it is not the goal of the nuts package, a note mentioning which package(s) (or software) was(were) used to create those maps could be of interest for the reader.

#### Functionality

The function nuts_test_multiple_versions do not have an examples. Would it be possible to have at least one?

After a nuts_convert_version(), it might be interesting to have a helping function that computes the estimated variations of the values between the different NUTS versions. It could be the difference (increase or decrease) in absolute or relative frequency at the higher NUT level. Like to get the 7.7% that you estimated in your first manure example (Spatial interpolation in a nutshell)

Hope it helps, 

Best,
Nolwenn

**Comment by krausewe** [comment]
Amazing, @nolwenn. Thank you for the feedback!

(And sorry for already starting to revise the package while you were still reviewing.)

**Comment by maelle** [comment]
Thanks a lot @nolwenn for your review! :pray:

**Comment by maelle** [comment]
@ropensci-review-bot submit review https://github.com/ropensci/software-review/issues/623#issuecomment-1961501137 time 2

**Comment by maelle** [comment]
@AAoritz @krausewe both reviews are in! :tada:

**Comment by maelle** [comment]
@AAoritz @krausewe thank you! Could you please record your response for the bot? https://devdevguide.netlify.app/bot_cheatsheet#submit-response-to-reviewers

@jospueyo @nolwenn could you please read the response of the authors and respond? Once you are happy with the changes, please use the approval template https://devdevguide.netlify.app/approval2template

**Comment by jospueyo** [comment]
## Reviewer Response

This has been awesome. I enjoyed reviewing your package and monitoring its evolution. I hope you are truly happy with my suggestions, I am with your responses.

Regarding `.by` argument, I didn't know it was still experimental. Otherwise, I wouldn't recommend the change.

I also agree with your opinion about imputation of missing values.

#### Final approval (post-review)

- [X] **The author has responded to my review and made changes to my satisfaction. I recommend approving this package.**



Estimated hours spent reviewing: 3.5

**Comment by AAoritz** [comment]
@ropensci-review-bot submit response <https://github.com/ropensci/software-review/issues/623#issuecomment-1976920666>

**Comment by maelle** [comment]
@nolwenn could you please read the response of the authors and respond?

**Comment by nolwenn** [comment]
## Reviewer Response

I enjoyed reviewing your package and was happy you found my comments valuable.  Sorry for the late replies to your answers. 

About the remaining points of discussion:
1.  I have no longer the namespace error message. All seems working out well.

2. `r nuts_convert_version()`
- including an option in `r nuts_convert_version()` that could return e.g. a column in the converted dataset containing more details on the conversion could indeed be an interesting way of doing implementing it.
- an example in the vignette on how to  extracted, by sub-setting and transforming the tibble data(cross_walks) is also nice

3. Missing regions : weights missing due to missing regions by changing the `r missing_rm` option to a threshold at which the users feel confident to assume 0 for the missing regions is a good idea to implement.

#### Final approval (post-review)

- [ x]  **The author has responded to my review and made changes to my satisfaction. I recommend approving this package.**



Estimated hours spent reviewing: 1 hour

**Comment by AAoritz** [comment]
One question on verbosity control @maelle. We thought that our current implementation using `rlang::local_options()` based on your recommendations would set verbosity as a default. However, verbosity has to be switched on to function. Do you understand why?

**Comment by maelle** [comment]
In 
https://github.com/AAoritz/nuts/blob/cf42cce8e522d4bd9ae1e8506024620dd964cebe/R/nuts_convert_version.R#L299

```r
is_verbose_mode <- (getOption("nuts.verbose", "quiet") == "verbose")
```

The default is "quiet", not "verbose".

**Comment by maelle** [comment]
Thanks @jospueyo & @nolwenn for your work as reviewers!

Thanks @AAoritz @krausewe for your work on your package!

**Comment by maelle** [comment]
@ropensci-review-bot approve nuts

**Comment by AAoritz** [comment]
Makes sense... we set the default to `verbose`. Thank you @maelle for this, your super helpful comments and your work as the editor during the review process! @krausewe and I learned a lot!

**Comment by AAoritz** [comment]
@nolwenn and @jospueyo, many thanks again! Please let us know if we can add you in the meta description as reviewers of the package. @maelle can we include you as well?!

**Comment by AAoritz** [comment]
@ropensci-review-bot finalize transfer of https://github.com/ropensci/nuts

**Comment by maelle** [comment]
> @maelle can we include you as well?!

Thanks for asking, but there's no need to list editors. Your participation in this process is thanks enough :smile_cat:

**Comment by jospueyo** [comment]
> > @nolwenn and @jospueyo, many thanks again! Please let us know if we can add you in the meta description as reviewers of the package. @maelle can we include you as well?!
> 
> Thanks for proposing. I do not know the convention but do as you like. As @maelle I am already happy to know that I helped.

I'd be honoured to be recognized for my tiny contribution to your so useful package.
